[BE] refactor: 비회원 Anonymous 처리 추가, Kakao ErrorHandling 개선, OAuth 추상화 #338

gabean13 opened the pull request:
<!-- PR 제목 예시
ex. [BE] 로그인 API 구현
ex. [AN] 페이지 구현 -->

## 📌 변경 내용 & 이유
<!-- ex. 
- 사용자 로그인 기능 구현
- 로그인 실패 시 에러 메시지 반환 처리 추가 -->

swagger 불필요한 cors 제거  [[BE] refactor: cors 제거](https://github.com/woowacourse-teams/2025-hEARit/commit/2e6895e2289ca7b07a97c5d207e2cb499f316ba2)

비회원 anonymous 처리 추가 [[BE] feat: 비회원 anonymous 처리 추가](https://github.com/woowacourse-teams/2025-hEARit/commit/4781ca82b0a59ba6e8b9b4de22b4cca6cb02de93) [[BE] refactor: member == null 제거, 비회원 체크 추가](https://github.com/woowacourse-teams/2025-hEARit/commit/dd7bbd110bd4b339052cd8d2f26952880ee7aa29) [[BE] rename: UserContext 패키지 domain으로 이동](https://github.com/woowacourse-teams/2025-hEARit/commit/c98fb6e1916c1f7aee10eda4e35c365dcfc77f06)
- CurrentMember -> UserContext로 클래스 네이밍 변경 
- WhiteList인데, Token 비어있으면 UserContext에 memberId는 null이고, isAuthenticated는 false인 객체 할당
- Service, Controller에서 CurrentMember == null이면 비회원 처리하도록 한 코드 수정

 추후에 Google, Naver등 다른 OAuth 추가되더라도 AuthService Layer 변경되지 않도록 리팩토링 [[BE] refactor: OAuth Service Layer 추가](https://github.com/woowacourse-teams/2025-hEARit/commit/f0f230a9bf624d39f4f1d344f2d9de327f912394)
[[BE] refactor: KaKao 관련 Class가 OAuth interface 구현하도록 변경](https://github.com/woowacourse-teams/2025-hEARit/commit/62f368824bcae9143d41f9bbc06d6d4e57d86e1b)
- OAuthService (interface) : 로그인을 위해 실제로 구현해야하는 애
- OAuthServiceRegistry (provider를 알면 oauthservcie구현체로 mapping해주는 객체)
- OAuthProvider (enum) : ex kakao, google .. 

Kakao Error Handling 하는 class 구현 & KaKao Timeout 추가 [[BE] feat: KaKao ErrorHandler 및 timeout 핸들링 추가](https://github.com/woowacourse-teams/2025-hEARit/commit/65a8acf3ce6181328fc2a5b9a9aa5593cc2513e2)
- ErrorCode 추가 : OAuth 관련 에러
- KakaoErrorHandler : 요청 보낼 때 에러 터지면 얘가 Handling함, 401만 특수하게 처리하고 나머지는 internal server error, kakao가 보내는 에러를 KaKaoException(HearitException 상속)로 변경해서 다시 throw함

## 📢 논의하고 싶은 내용
<!-- 이러이러해서 이부분을 이러이러하게 구현했는데 이방식이 괜찮은지? -->

Kakao Error랑 OAuth 추상화 부분 집중해서 봐주시면 좋을듯염

두 부분 집중해서 봐주세용

## 🧩 관련 이슈
<!-- 이슈 태그: #123 -->
<!-- 이슈 닫기: close #123 -->

close #338 

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- 신기능
  - OAuth 기반 소셜 로그인 통합(카카오 포함).
  - 게스트 컨텍스트 도입으로 탐색/상세/추천에서 회원·비회원별 응답 제공(로그인 시 북마크 상태 포함).
  - 추천 카테고리 및 탐색 개인화 개선.
- 리팩터링
  - 인증 주체를 UserContext로 통일.
  - 공개 GET 엔드포인트 정리, 나머지 인증 요구.
- 테스트
  - OAuth/보안/서비스 흐름 반영해 테스트 전면 업데이트.
- 잡무
  - 로그인/회원가입/토큰 재발급 요청 검증 강화.
  - OAuth/Kakao 클라이언트 설정 및 타임아웃 추가.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

gabean13 committed:
[BE] refactor: cors 제거 #338

gabean13 committed:
[BE] feat: 비회원 anonymous 처리 추가 #338

- WhiteList인데 Authenticate안 된 경우 UserContext guest로 할당

gabean13 committed:
[BE] refactor: member == null 제거, 비회원 체크 추가 #338

gabean13 committed:
[BE] refactor: OAuth Service Layer 추가 #338

gabean13 committed:
[BE] rename: UserContext 패키지 domain으로 이동 #338

gabean13 committed:
[BE] refactor: KaKao 관련 Class가 OAuth interface 구현하도록 변경 #338

gabean13 committed:
[BE] feat: KaKao ErrorHandler 및 timeout 핸들링 추가 #338

gabean13 committed:
[BE] rename: KaKao Error 관련 클래스 exception 패키지로 이동 #338

gabean13 commented:
@coderabbitai 레빗아 리뷰달아줘

gabean13 commented:
@coderabbitai 레빗아 리뷰달으라고

gabean13 committed:
[BE] refactor: userContext null check 추가 #338

gabean13 committed:
[BE] refactor: cursorRequest 추가 및 검증 #338

gabean13 committed:
[BE] feat: Member OAuthProvider Column 추가 #338

gabean13 committed:
[BE] refactor: GET만 허용하도록 API 화이트 리스트 수정 #338

gabean13 commented on backend/src/main/java/com/onair/hearit/auth/infrastructure/oauth/kakao/KakaoClientConfig.java:
```diff
@@ -0,0 +1,36 @@
+    private final KakaoClientProperties kakaoClientProperties;
+    private final ObjectMapper objectMapper;
+
+    @Bean
+    public RestClient kakaoRestClient() {
+        return RestClient.builder()
+                .requestFactory(getClientHttpRequestFactory())
+                .defaultStatusHandler(new KakaoErrorHandler(objectMapper))
+                .baseUrl(kakaoClientProperties.getBaseUrl())
+                .build();
+    }
```
근데 짜피 그렇게 들어가잔슴 내가 그렇게 설정안해줘도 메서드명으로 들어가잔아

gabean13 committed:
[BE] refactor: UserContext memberId null 막음 #338

gabean13 committed:
[BE] refactor: Kakao timeout int -> Duration으로 변경 #338

gabean13 committed:
[BE] refactor: OAuthError는 BADGATEWAY로 변경 #338

JO-eusan commented on backend/src/main/java/com/onair/hearit/application/HearitService.java:
```diff
@@ -78,19 +87,42 @@ public List<HearitsWithRecommendCategoryResponse> getHearitsWithRecommendCategor
         List<Long> categoryIds = getAllCategoryIdsWithoutRecommend(recommendCategories);
-        Collections.shuffle(categoryIds, new Random(seed));
-        return categoryIds.subList(0, count);
+        List<Long> mutableCategoryIds = new ArrayList<>(categoryIds);
```
mutable으로 변경하신 이유가 있나요?

gabean13 replied:
이거 getAllCategorytIdsWithoutRecommend()의 반환 값이 immutable이라서 내부에서 섞을 때 새로운 변수에 할당해서 섞고 자른거였어영
그 반환값을 변하는 걸로 해도 되긴 하는데 (기존) , 반환 값을 변경하는 것 보다, 섞는 곳에서 새로운 변수에 할당하는게 더 나을 것 같아서 이렇게 바꿨으무

yuyoungrhee commented on backend/src/main/java/com/onair/hearit/auth/infrastructure/jwt/JwtAuthenticationFilter.java:
```diff
@@ -50,9 +51,10 @@ protected void doFilterInternal(HttpServletRequest request, HttpServletResponse
         }
 
```
래빗이가 UserContext.guest() 설정 그냥 Jwt Filter에서 명시적으로 하라는데 어케생각함요?

멍PT
- @AuthenticationPrincipal UserContext는 anonymous principal 설정 덕분에 잘 들어오는 건 맞음(현재 문제 없음)
- 인증 흐름의 제어권을 명확히 갖고 싶다면 JwtAuthenticationFilter에서 설정하는 것이 바람직
- 결국 “누가 guest라고 판단할 것인가?“를 Spring이 아닌 너희 앱이 제어할 것인가가 핵심

```java
        if ((token == null || token.isBlank()) && isWhitelisted(request)) {
            UsernamePasswordAuthenticationToken auth =
                    new UsernamePasswordAuthenticationToken(UserContext.guest(), null, null);
            SecurityContextHolder.getContext().setAuthentication(auth);
            chain.doFilter(request, response);
            return;
        }
```

gabean13 replied:
둘다 하는게 좋을 것 같아서 둘다 하겟음둥

JO-eusan commented on backend/src/main/java/com/onair/hearit/auth/application/OAuthProvider.java:
```diff
@@ -0,0 +1,9 @@
+package com.onair.hearit.auth.application;
+
+public enum OAuthProvider {
```
이 이넘 클래스는 도메인에 두는건 어떤가요?
어떤 서비스 로직이 들어 있는 것이 아니라, 우리가 사용하는 인증 도메인을 작성하는 클래스라고 생각합니다.

Byesol replied:
저도 해당 사용자가 '카카오 유저'인지 '구글 유저'인지를 구분하는 건 핵심 비즈니스 로직 같아서 도메인 패키지에 어울리는 클래스라고 생각해요

gabean13 replied:
아하 좋아염!

JO-eusan commented on backend/src/main/java/com/onair/hearit/auth/dto/request/OAuthLoginRequest.java:
```diff
@@ -0,0 +1,9 @@
+import jakarta.validation.constraints.NotBlank;
+
+public record OAuthLoginRequest(
+        @NotBlank
```
혹시 `@NotBlank String accessToken`이렇게 작성하는건 별론가여. 컨벤션 논의 한번 해봐용

JO-eusan commented on backend/src/main/java/com/onair/hearit/common/exception/ErrorCode.java:
```diff
@@ -12,7 +12,7 @@ public enum ErrorCode {
     METHOD_NOT_ALLOWED(HttpStatus.METHOD_NOT_ALLOWED, "지원하지 않는 HTTP 메서드입니다."),
     INTERNAL_SERVER_ERROR(HttpStatus.INTERNAL_SERVER_ERROR, "서버 내부 오류입니다."),
-    ;
+    OAUTH_CLIENT_ERROR(HttpStatus.BAD_GATEWAY, "OAuth 관련 에러입니다.");
```
인증 예외랑 가깝게 두는건 어떤가요?

gabean13 replied:
근데 이거 503이라서 internal server error 아래에 잇는게 맞다고 생각했어욤

JO-eusan commented on backend/src/main/resources/application-dev.properties:
```diff
@@ -1,52 +1,42 @@
 # secret import
 spring.config.import=optional:file:.env[.properties]
-
 # Dev-DataSource
```
요 개행 없어진거 의도한건가여

Byesol replied:
저도 이 부분 건들때마다  저장 시 개행 없어지더라고요.

아예 이 파일은 CODE STYLE 적용 빼니까 해결됐어요

gabean13 replied:
ㅠㅠ 개행 다시 살리고 저도 뺄게유

JO-eusan commented on backend/src/test/resources/application-fake-test.properties:
```diff
@@ -2,25 +2,23 @@ spring.datasource.url=jdbc:h2:mem:testdb;DB_CLOSE_DELAY=-1;DB_CLOSE_ON_EXIT=FALS
 spring.datasource.driverClassName=org.h2.Driver
 spring.datasource.username=sa
 spring.datasource.password=
```
여기도 혹시 의도한건가요? 개행 있는 버전이 더 가독성 좋지 않을까용

gabean13 replied:
개행 다시 추가 햇어요

JO-eusan commented on backend/src/main/java/com/onair/hearit/application/explore/HearitExploreService.java:
```diff
@@ -37,56 +44,78 @@ public class HearitExploreService {
-        if (memberId == null) {
-            return exploredHearitQueryRepository.findExploredHearitsForGuest(cursorId, size);
+    private List<Hearit> getExploredHearitsForGuest(Long cursorId, int size) {
+        Long guestId = -1L;
```
여기 CurrentMember에 비회원이면 -1로 나오는데 이렇게 하드코딩 안해도 되지 않을까요?

gabean13 replied:
근데 그건 null처리 안하려고 -1이라는 값을 넣은건데, 
보통 비회원이면 -1인걸 이용해서 데이터베이스에 넣는게 일반적인 상황은 아닌 것 같아서 관련 코드 상에서 보여야한다고 생각해서 이렇게 했어욥

JO-eusan reviewed (changes requested):
가콩 변경된 부분 확인했습니다. 커멘트 확인 부탁드려요.

Byesol commented on backend/src/test/java/com/onair/hearit/application/HearitExploreServiceTest.java:
```diff
@@ -88,52 +96,57 @@ void generatePersonalScore() {
-        Hearit hearit7 = dbHelper.insertHearit(createHearitByNameAndCategory("hearit7", category2));
-        Hearit hearit8 = dbHelper.insertHearit(createHearitByNameAndCategory("hearit8", category3));
-        Hearit hearit9 = dbHelper.insertHearit(createHearitByNameAndCategory("hearit9", category3));
-        Hearit hearit10 = dbHelper.insertHearit(createHearitByNameAndCategory("hearit10", category3));
```
테스트의 가독성을 위해 로컬 지역 변수를 굳이 사용하지 않더라도 저는 선언하는 게 더 좋다고 생각하는데 이에 대해서 어떻게 생각하시나요?

gabean13 replied:
저는 안 쓸건데 지역변수에 저장해서 메모리를 소모할 이유가 없다고 생각했어요 
의미는 insertHearit으로 충분히 드러나지 않나요??? 
그리고 오히려 안쓰는데 변수를 선언해놔서 더 가독성이 떨어진다고 생각했어요 처음에 봤을 때 이거 왜 선언한거지 하는 생각이 먼저 들더라구여

Byesol commented on backend/src/main/java/com/onair/hearit/application/BookmarkService.java:
```diff
@@ -53,20 +53,20 @@ public BookmarkInfoResponse addBookmark(CurrentMember currentMember, Long hearit
-    private Member getMemberByCurrentMember(CurrentMember currentMember) {
-        if (currentMember == null) {
+    private Member getMemberByUserContext(UserContext userContext) {
+        if (userContext == null || userContext.isGuest()) {
```
비회원 anoymous 처리 한거면 이제 정상 흐름 내에서는 userContext가 null 인 경우가 없나요?
그렇다면 아래 테스트에서 userContext를 null로 주고 있던데 아래 테스트코드를 고쳐야할 것 같아요
```
    @DisplayName("로그인하지 않은 사용자는 북마크 목록 조회시 Unauthorized 예외를 던진다.")
    void getBookmarkHearitsTest_
```

gabean13 replied:
ㄷㄷ 천재?!?! 할게염

Byesol commented on backend/src/main/java/com/onair/hearit/auth/application/AuthService.java:
```diff
@@ -90,7 +89,6 @@ private LoginTokenResponse createTokenResponseFrom(Member member) {
                         existing -> existing.update(refreshToken, expiryDate),
```
살짝 이번 리팩 범위랑 별개일 수도 있긴 한데, 인텔리제이에서 보다 보니 눈에 띄고,  변경한다면 해당 파일 변경하는 김에 변경하는 게 좋을 것 같아 말씀드리는데

혹시 지금 comment 단 부분인 아래 메서드를 

```java
    private LoginTokenResponse createTokenResponseFrom(Member member) {
        String accessToken = jwtTokenProvider.createAccessToken(member.getId());
        String refreshToken = jwtTokenProvider.createRefreshToken(member.getId());
        LocalDateTime expiryDate = jwtTokenProvider.extractExpiry(refreshToken);
        refreshTokenRepository.findByMemberId(member.getId())
                .ifPresentOrElse(
                        existing -> existing.update(refreshToken, expiryDate),
                        () -> refreshTokenRepository.save(new RefreshToken(member.getId(), refreshToken, expiryDate))
                );
        return new LoginTokenResponse(accessToken, refreshToken);
    }
```

```java

private LoginTokenResponse createTokenResponseFrom(Member member) {
    String accessToken = jwtTokenProvider.createAccessToken(member.getId());
    String refreshToken = jwtTokenProvider.createRefreshToken(member.getId());

    saveOrUpdateRefreshToken(member, refreshToken); 

    return new LoginTokenResponse(accessToken, refreshToken);
}

private void saveOrUpdateRefreshToken(Member member, String refreshToken) {
    LocalDateTime expiryDate = jwtTokenProvider.extractExpiry(refreshToken);
    refreshTokenRepository.findByMemberId(member.getId())
            .ifPresentOrElse(
                    existing -> existing.update(refreshToken, expiryDate),
                    () -> refreshTokenRepository.save(new RefreshToken(member.getId(), refreshToken, expiryDate))
            );
}
```
이런 식으로 변경해주실 수 있나요? 같이 보니 가독성 측면에서 해당 메서드가 눈에 안 들어와서요.

gabean13 replied:
오 조은데요 햇어영

Byesol commented on backend/src/main/java/com/onair/hearit/auth/infrastructure/oauth/kakao/KakaoOAuthService.java:
```diff
@@ -0,0 +1,35 @@
+    @Override
+    public OAuthUserInfo fetchUser(String accessToken) {
+        KakaoUserInfoResponse response = getUserInfo(accessToken);
+        return new OAuthUserInfo(response.id(), response.nickname(), response.profileImage(), OAuthProvider.KAKAO);
```
여기서도 그냥 위의  provider 메서드를 통해 OAuthProvider.KAKAO 를 제공해도 되지 않을까요?

gabean13 replied:
오 대박쓰 햇어요

Byesol commented on backend/src/main/java/com/onair/hearit/auth/application/AuthService.java:
```diff
@@ -66,18 +65,18 @@ private void validateDuplicatedId(SignupRequest request) {
     }
 
-    // 소셜 로그인의 경우 회원가입이 따로 없으며 로그인 시 자동으로 회원가입되도록 구현
     @Transactional
```
이 메서드의 경우 외부 연결이 메서드 내에 포함되어 있어서 이 메서드에 붙은 transactional을 아래의

```
        Member member = memberRepository.findBySocialIdAndOAuthProvider(userInfo.id(), provider)
                .orElseGet(() -> signupWithUserInfo(userInfo, provider));
```

부분만 메서드 분리해서 여기에만 transactional 붙이는 건 어떨까요?

gabean13 replied:
근데 그러면 이 메서드의 전체 흐름이 끝까지 일어나야지 정상적인 흐름인건데 member만 Transatcional을 하면 해당 부분 이후에 흐름이 실패하면 되돌리지 못하는거아니에요??

Byesol reviewed (commented):
정말 수고많으셨어요! 몇 가지 커멘트 남길께요

gabean13 committed:
[BE] rename: OAuthProvider domain으로 패키지 이동 #338

gabean13 committed:
[BE] style: .properties 파일 개행 추가 #338

gabean13 committed:
[BE] refactor: 메서드 분리 및 테스트 코드 가독성 개선 #338

gabean13 committed:
[BE] refactor: request validation 추가 #338

yuyoungrhee reviewed (commented):
가콩 의견 하나만 확인해주세욘

gabean13 committed:
[BE] Merge branch 'develop-be' into refactor/#338-auth-refactoring
#338

# Conflicts:
#	backend/src/main/java/com/onair/hearit/auth/dto/request/KakaoLoginRequest.java
#	backend/src/main/java/com/onair/hearit/auth/dto/request/LoginRequest.java
#	backend/src/main/java/com/onair/hearit/auth/dto/request/SignupRequest.java
#	backend/src/main/java/com/onair/hearit/auth/dto/request/TokenReissueRequest.java

gabean13 committed:
[BE] test: test 제거 #338

yuyoungrhee reviewed (approved):
확인했습니다!!! 수고많았어용 가콩

JO-eusan reviewed (approved):
코드 변경 부분 확인했습니다!
연휴 동안 고생많으셨어요 가콩 (-_-👍🏻)
